app/controllers/idv/scan_id_controller.rb

```diff
@@ -0,0 +1,273 @@
+      return unless data
+      scan_id_session[:facematch_pass] = facematch_pass?(data)
+    end
+
+    def facematch_body(image1, image2)
+      { 'Data':
+        { 'ImageOne': image1,
+          'ImageTwo': image2 },
+        'Settings':
+          { 'SubscriptionId': Figaro.env.acuant_assure_id_subscription_id } }.to_json
+    end
+
+    def fetch_face_from_document
+      assure_id.instance_id = scan_id_session[:instance_id]
+      data = proxy_request { assure_id.face_image }
+      return if data.blank?
+      Base64.strict_encode64(data)
+    end
+
+    def process_selfie
+      liveness_body = request.body.read
+      liveness_data = proxy_request { liveness_service.liveness(liveness_body) }
+      return unless liveness_data
+      unless selfie_live?(liveness_data)
+        render_json({})
+        return
+      end
+      JSON.parse(liveness_body)['Image']
+    end
+
+    def process_data_and_store_pii_in_session_if_document_passes(data, instance_id)
+      data = JSON.parse(data)
+      if data['Result'] == ACUANT_PASS
+        scan_id_session[:instance_id] = instance_id
+        scan_id_session[:pii] =
+          Idv::Utils::PiiFromDoc.new(data).call(current_user&.phone_configurations&.take&.phone)
+      end
+      data
+    end
+
+    def facematch_pass?(data)
+      JSON.parse(data)['IsMatch']
+    end
+
+    def selfie_live?(data)
+      JSON.parse(data)['LivenessResult']['LivenessAssessment'] == 'Live'
+    end
+
+    def flow_session
+      user_session[USER_SESSION_FLOW_ID]
+    end
+
+    def assure_id
+      @assure_id ||= new_assure_id
+    end
+
+    def new_assure_id
+      (Rails.env.test? ? Idv::Acuant::FakeAssureId : Idv::Acuant::AssureId).new
+    end
+
+    def liveness_service
+      (Rails.env.test? ? Idv::Acuant::FakeAssureId : Idv::Acuant::Liveness).new
+    end
+
+    def facematch_service
+      (Rails.env.test? ? Idv::Acuant::FakeAssureId : Idv::Acuant::FacialMatch).new
+    end
+
+    def ensure_fully_authenticated_user_or_token
+      return if user_signed_in? && user_fully_authenticated?
+      ensure_user_id_in_session
+    end
+
+    def ensure_user_id_in_session
+      return if token_user_id && token.blank?
+      result = CaptureDoc::ValidateRequestToken.new(token).call
+      analytics.track_event(Analytics::DOC_AUTH, result.to_h)
+      process_result(result)
+    end
+
+    def process_result(result)
+      if result.success?
+        reset_session
+        session[:token_user_id] = result.extra[:for_user_id]
+      else
+        flash[:error] = t('errors.capture_doc.invalid_link')
+        redirect_to root_url
+      end
+    end
+
+    def all_checks_passed?
+      scan_id_session && scan_id_session[:instance_id] && scan_id_session[:facematch_pass] &&
+        (scan_id_session[:liveness_pass] || !FeatureManagement.liveness_checking_enabled?)
+    end
+
+    def token
+      params[:token]
+    end
+
+    def render_json(data)
```

Comment:
Do we need this helper method? how often is `data` nil?

Reply:
The proxied requests can be nil.  Since the universe is shying away from throwing exceptions it was a way of handling the proxy requests dying with service unavailable.